File: part-2-simple-ut-particle/cpp/test-physics-correction.cpp
/*********************************************************/
#include "physics.hpp"
#include <gtest/gtest.h>

/*********************************************************/
using namespace experiment;

/*********************************************************/
TEST(TestPhysics, collide)
{
	//build
	Particle particle1(-1.0, 1.0);
	Particle particle2(1.0, -1.0);

	//call
	EXPECT_FALSE(physics::collide(particle1, particle2, 0.5)) << "A cause d'une erreur";
	EXPECT_FALSE(physics::collide(particle1, particle2, 0.9));
	EXPECT_TRUE(physics::collide(particle1, particle2, 1.0));
	EXPECT_TRUE(physics::collide(particle1, particle2, 2.0));
}

/*********************************************************/
TEST(TestPhysics, get_collision_time_1_symetric)
{
	//build
	Particle particle1(-1.0, 1.0);
	Particle particle2(1.0, -1.0);

	//call
	EXPECT_EQ(1.0, physics::get_collision_time(particle1, particle2));
}

/*********************************************************/
TEST(TestPhysics, get_collision_time_2_asymetric)
{
	//build
	Particle particle1(-0.5, 1.0);
	Particle particle2(1.0, -1.0);

	//call
	EXPECT_EQ(0.75, physics::get_collision_time(particle1, particle2));
}

/*********************************************************/
TEST(TestPhysics, test_elastic_collision_1_symetric)
{
	//build
	Particle particle1(-1.0, 1.0);
	Particle particle2(1.0, -1.0);

	//call
	bool ret = physics::elastic_collision(particle1, particle2, 2);
	ASSERT_TRUE(ret);

	//check states
	EXPECT_EQ(-1.0, particle1.get_x());
	EXPECT_EQ(-1.0, particle1.get_vx());
	EXPECT_EQ(1.0, particle2.get_x());
	EXPECT_EQ(1.0, particle2.get_vx());
}

File: part-2-simple-ut-particle/python/test_physics-correction.py
from unittest import TestCase
from particle import Particle
import physics

class TestPysiscs(TestCase):
	def test_collide(self):
		particle1 = Particle(-1.0, 1.0)
		particle2 = Particle(1.0, -1.0)
		self.assertFalse(physics.collide(particle1, particle2, 0.5))
		self.assertFalse(physics.collide(particle1, particle2, 0.9))
		self.assertTrue(physics.collide(particle1, particle2, 1.0))
		self.assertTrue(physics.collide(particle1, particle2, 2.0))
	
	def test_get_collision_time_1_symetric(self):
		particle1 = Particle(-1.0, 1.0)
		particle2 = Particle(1.0, -1.0)
		self.assertEqual(1.0, physics.get_collision_time(particle1, particle2))

	def test_get_collision_time_2_asymetric(self):
		particle1 = Particle(-0.5, 1.0)
		particle2 = Particle(1.0, -1.0)
		self.assertEqual(0.75, physics.get_collision_time(particle1, particle2))

	def test_elastic_collision_1_symetric(self):
		particle1 = Particle(-1.0, 1.0)
		particle2 = Particle(1.0, -1.0)
		ret = physics.elastic_collision(particle1, particle2, 2)
		self.assertTrue(ret)
		self.assertEqual(-1.0, particle1.get_x())
		self.assertEqual(-1.0, particle1.get_vx())
		self.assertEqual(1.0, particle2.get_x())
		self.assertEqual(1.0, particle2.get_vx())